src/uct/cuda/base/cuda_md.c

```diff
@@ -184,75 +237,44 @@ UCS_PROFILE_FUNC(ucs_status_t, uct_cuda_base_mem_query,
         return UCS_OK;
     }
 
-    if (address == NULL) {
-        mem_type              = UCS_MEMORY_TYPE_HOST;
-    } else {
-        attr_type[0] = CU_POINTER_ATTRIBUTE_MEMORY_TYPE;
-        attr_data[0] = &cuda_mem_mype;
-        attr_type[1] = CU_POINTER_ATTRIBUTE_IS_MANAGED;
-        attr_data[1] = &is_managed;
-        attr_type[2] = CU_POINTER_ATTRIBUTE_DEVICE_ORDINAL;
-        attr_data[2] = &cuda_device;
-
-        cu_err = cuPointerGetAttributes(ucs_static_array_size(attr_data),
-                                        attr_type, attr_data,
-                                        (CUdeviceptr)address);
-        if ((cu_err != CUDA_SUCCESS) || (cuda_mem_mype != CU_MEMORYTYPE_DEVICE)) {
-            /* pointer not recognized */
-            return UCS_ERR_INVALID_ADDR;
-        }
-
-        if (is_managed) {
-            mem_type = UCS_MEMORY_TYPE_CUDA_MANAGED;
-        } else {
-            mem_type = UCS_MEMORY_TYPE_CUDA;
-
-            /* Synchronize for DMA */
-            cu_err = cuPointerSetAttribute(&value,
-                                           CU_POINTER_ATTRIBUTE_SYNC_MEMOPS,
-                                           (CUdeviceptr)address);
-            if (cu_err != CUDA_SUCCESS) {
-                cuGetErrorString(cu_err, &cu_err_str);
-                ucs_warn("cuPointerSetAttribute(%p) error: %s", address,
-                         cu_err_str);
-            }
-        }
-
-        if (mem_attr->field_mask & UCT_MD_MEM_ATTR_FIELD_SYS_DEV) {
-            status = uct_cuda_base_get_sys_dev(cuda_device, &sys_dev);
+    if (address != NULL) {
+        status = ucs_memtype_cache_lookup(address, length, &addr_mem_info);
```

Comment:
looks like it would result it 2 cache lookups (UCP, then UCT)
maybe remove memtype cache lookup from this function, and use it from some other place that needs to detect the memory type for selecting stream index?

Reply:
removed lookup in UCT md query function. Now, before checking any mds, we first do a lookup. If absent, we ask the mds to query and the mds update global cache. 

Later if UCT itself wants to query memory attributes, it does a cache lookup and if that fails, it does an internal query. (This part is out of scope for this PR and needs to go to the PR that does stream selection based on memory types of src/dst buffers)